x/emissions/module/rewards_v2.go

```diff
@@ -0,0 +1,430 @@
+	stakeWeightedLoss := make([]float64, len(reputersReportedLosses[0]))
+	for j := 0; j < len(reputersReportedLosses[0]); j++ {
+		logSum := 0.0
+		for i, losses := range reputersReportedLosses {
+			logSum += (math.Log10(losses[j]) * reputersAdjustedStakes[i]) / totalStake
+		}
+		stakeWeightedLoss[j] = logSum
+	}
+
+	return stakeWeightedLoss, nil
+}
+
+// GetConsensusScore calculates the proximity to consensus score for a reputer.
+// T_im
+func GetConsensusScore(reputerLosses, consensusLosses []float64) (float64, error) {
+	fTolerance := 0.01
+	if len(reputerLosses) != len(consensusLosses) {
+		return 0, types.ErrInvalidSliceLength
+	}
+
+	var sumLogConsensusSquared float64
+	for _, cLoss := range consensusLosses {
+		sumLogConsensusSquared += math.Pow(cLoss, 2) //!
+	}
+	consensusNorm := math.Sqrt(sumLogConsensusSquared)
+
+	var distanceSquared float64
+	for i, rLoss := range reputerLosses {
+		distanceSquared += math.Pow(math.Log10(rLoss/consensusLosses[i]), 2)
+	}
+	distance := math.Sqrt(distanceSquared)
+
+	score := 1 / (distance/consensusNorm + fTolerance)
+	return score, nil
+}
+
+// GetAllConsensusScores calculates the proximity to consensus score for all reputers.
+// calculates:
+// T_i - stake weighted total consensus
+// returns:
+// T_im - reputer score (proximity to consensus)
+func GetAllConsensusScores(allLosses [][]float64, stakes []float64, allListeningCoefficients []float64, numReputers int) ([]float64, error) {
+	// Get adjusted stakes
+	var adjustedStakes []float64
+	for i, reputerStake := range stakes {
+		adjustedStake, err := GetAdjustedStake(reputerStake, stakes, allListeningCoefficients[i], allListeningCoefficients, float64(numReputers))
+		if err != nil {
+			return nil, err
+		}
+		adjustedStakes = append(adjustedStakes, adjustedStake)
+	}
+
+	// Get consensus loss vector
+	consensus, err := GetStakeWeightedLossMatrix(adjustedStakes, allLosses)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get reputers scores
+	scores := make([]float64, numReputers)
+	for i := 0; i < numReputers; i++ {
+		losses := allLosses[i]
+		scores[i], err = GetConsensusScore(losses, consensus)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return scores, nil
+}
+
+// GetAllReputersOutput calculates the final scores and adjusted listening coefficients for all reputers.
+// This function iteratively adjusts the listening coefficients based on a gradient descent method to minimize
+// the difference between each reputer's losses and the consensus losses, taking into account each reputer's stake.
+// returns:
+// T_im - reputer score (proximity to consensus)
+// a_im - listening coefficients
+func GetAllReputersOutput(allLosses [][]float64, stakes []float64, initialCoefficients []float64, numReputers int, enableListening bool) ([]float64, []float64, error) {
+	learningRate := 0.01
+	coefficients := make([]float64, len(initialCoefficients))
+	copy(coefficients, initialCoefficients)
+
+	if enableListening {
```

Comment:
think instead of this control flow we agree to do it or not... there was control flow in the simulation cuz they were testing what happens with/without listening enabled. you can double check w diederik but quite sure the resolution was that it was needed => can remove this if-statement